peer/sync.go

```diff
@@ -0,0 +1,152 @@
+package peer
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"sync"
+
+	"github.com/renproject/aw/channel"
+	"github.com/renproject/aw/transport"
+	"github.com/renproject/aw/wire"
+	"github.com/renproject/id"
+	"go.uber.org/zap"
+)
+
+type pendingContent struct {
+	// content is nil while synchronisation is happening. After synchronisation
+	// has completed, content will be set.
+	content []byte
+
+	// cond is used to wait and notify goroutines about the completion of
+	// synchronisation.
+	cond *sync.Cond
+}
+
+// wait for the content to be synchronised. Returns a channel that the caller
+// can block on while waiting for the content. This is better than blocking
+// interally, because it can be composed with contexts/timeouts.
+func (pending *pendingContent) wait() <-chan []byte {
+	w := make(chan []byte, 1)
+	go func() {
+		pending.cond.L.Lock()
+		for pending.content == nil {
+			pending.cond.Wait()
+		}
+		content := make([]byte, len(pending.content), len(pending.content))
+		copy(content, pending.content)
+		pending.cond.L.Unlock()
+		w <- content
+	}()
+	return w
+}
+
+// signal that the content is synchronised. All goroutines waiting on the
+// content will be awaken and will create their own copies of the content.
+func (pending *pendingContent) signal(content []byte) {
+	pending.cond.L.Lock()
+	pending.content = content
+	pending.cond.L.Unlock()
+}
+
+type Syncer struct {
+	opts      SyncerOptions
+	filter    *channel.SyncFilter
+	transport *transport.Transport
+
+	pendingMu *sync.Mutex
+	pending   map[string]pendingContent
+}
+
+func NewSyncer(opts SyncerOptions, filter *channel.SyncFilter, transport *transport.Transport) *Syncer {
+	return &Syncer{
+		opts:      opts,
+		filter:    filter,
+		transport: transport,
+
+		pendingMu: new(sync.Mutex),
+		pending:   make(map[string]pendingContent, 1024),
```

Comment:
I think this should be similar to the cache resolver, where it drops elements when the cache size is exceeded, and shouldn't grow unbounded

Reply:
It is not unbounded. Once a Sync is complete (successful or not) the respective entry in this map is deleted. We could restrict the number of pending Syncs happening at any one time, but I feel it is better to just have the user of Airwave restrict themselves.